### PR TITLE
Apache NiFi charm

### PR DESCRIPTION
**Pull request by mhmoerma**
This merge contains the new Apache NiFi charm, including it's trusty build. It also has the changes made to the Java layer.

### CODE REVIEW

File: charms/layers/apache-nifi/icon.svg [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+       style="fill:url(#linearGradient6461);fill-opacity:1;stroke:none;display:inline;filter:url(#filter1121)"
+       d="m -268,700.15563 0,-33.72973 c 0,-27.24324 3.88785,-31.13513 31.10302,-31.13513 l 33.79408,0 c 27.21507,0 31.1029,3.89189 31.1029,31.13513 l 0,33.72973 c 0,27.24325 -3.88783,31.13514 -31.1029,31.13514 l -33.79408,0 C -264.11215,731.29077 -268,727.39888 -268,700.15563 Z"
+       id="path6455"
+       inkscape:connector-curvature="0"
+       sodipodi:nodetypes="sssssssss" />
+    <image
+       y="653.25421"
+       x="-244.96342"
+       id="image3776"
+       xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADwAAABMCAYAAAAx3tSzAAAABHNCSVQICAgIfAhkiAAACkdJREFU
```
Comment (galgalesh) [inline charms/layers/apache-nifi/icon.svg]: This is a PNG image embedded in an svg image. Don't they have an svg logo somewhere?

File: charms/layers/apache-nifi/reactive/apache_nifi.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    tfile = tarfile.open(hookenv.resource_get('apache-nifi'), 'r')
+    filesdir = '%s/files' % hookenv.charm_dir()
+    tfile.extractall(filesdir)
+    re_edit_in_place('%s/nifi-1.0.0/conf/nifi.properties' % filesdir, {
+        r'.*nifi.web.http.port.*': 'nifi.web.http.port=%s' % conf['nifi-port'],
+    })
+    subprocess.check_call(['bash', '%s/nifi-1.0.0/bin/nifi.sh' % filesdir, 'install'])
+    service_restart('nifi')
+    hookenv.open_port(conf['nifi-port'])
+    hookenv.status_set('active', 'Apache NiFi has been installed and is running in standalone mode')
```
Comment (galgalesh) [inline charms/layers/apache-nifi/reactive/apache_nifi.py]: Try to keep these messages as short as possible. Example: `Ready (standalone)`. Non-blocking but just something to keep in mind in the future.

File: charms/layers/apache-nifi/reactive/apache_nifi.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    re_edit_in_place('%s/files/nifi-1.0.0/conf/nifi.properties' % hookenv.charm_dir(), {
+        r'.*nifi.cluster.is.node.*': 'nifi.cluster.is.node=true',
+        r'.*nifi.cluster.node.address.*': 'nifi.cluster.node.address=%s' % hookenv.unit_public_ip(),
+        r'.*nifi.web.http.port.*': 'nifi.web.http.port=%s' % conf['nifi-port'],
+        r'.*nifi.cluster.node.protocol.port.*': 'nifi.cluster.node.protocol.port=%s' % conf['cluster-port'],
+        r'.*nifi.zookeeper.connect.string.*': 'nifi.zookeeper.connect.string=%s' % zookeeper_servers_string
+    })
+    service_restart('nifi')
+    set_state('apache-nifi.cluster')
+    hookenv.status_set('active', 'Apache NiFi is running in a cluster with Apache Zookeeper')
```
Comment (galgalesh) [inline charms/layers/apache-nifi/reactive/apache_nifi.py]: `Ready (clustered with Zookeeper)`

File: charms/layers/apache-nifi/reactive/apache_nifi.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    zookeeper_servers_string = ''
+    filesdir = '%s/files' % hookenv.charm_dir()
+    for zk_unit in zookeeper.zookeepers():
+        zookeeper_servers_string += '%s:%s,' % (zk_unit['host'], zk_unit['port'])
+    if zookeeper_servers_string[:-1] not in open('%s/nifi-1.0.0/conf/nifi.properties' % filesdir).read():
+        hookenv.status_set('maintenance', 'Zookeeper has changed. Updating Apache NiFi settings and restarting')
+        re_edit_in_place('%s/nifi-1.0.0/conf/nifi.properties' % filesdir, {
+            r'.*nifi.zookeeper.connect.string.*': 'nifi.zookeeper.connect.string=%s' % zookeeper_servers_string[:-1]
+        })
+        service_restart('nifi')
```
Comment (galgalesh) [inline charms/layers/apache-nifi/reactive/apache_nifi.py]: It's best to check if restart succeeded. Don't fail silently!

File: charms/layers/apache-nifi/reactive/apache_nifi.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+@when('apache-nifi.cluster')
+@when_not('zookeeper.joined', 'zookeeper.ready')
+def zookeeper_removed():
+    hookenv.status_set('maintenance', 'Removing Apache NiFi from cluster')
+    re_edit_in_place('%s/files/nifi-1.0.0/conf/nifi.properties' % hookenv.charm_dir(), {
+        r'.*nifi.cluster.is.node.*': 'nifi.cluster.is.node=false'
+    })
+    service_stop('nifi')
+    remove_state('apache-nifi.cluster')
+    hookenv.status_set('waiting', 'Apache NiFi is configured for standalone mode, but not running')
```
Comment (galgalesh) [inline charms/layers/apache-nifi/reactive/apache_nifi.py]: A user reading this message should know what to do to get it running again: `Stopped (waiting for zookeeper)`

File: charms/layers/apache-nifi/templates/nifi.properties [hunk 6]
```diff
@@ -0,0 +1,196 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (galgalesh) [inline charms/layers/apache-nifi/templates/nifi.properties]: You don't use this file anymore, right? Remove it.

File: charms/layers/apache-nifi/icon.svg (hunk 1 above)
Comment (gvseghbr) [inline charms/layers/apache-nifi/icon.svg]: https://upload.wikimedia.org/wikipedia/commons/f/ff/Apache-nifi-logo.svg

**Comment by galgalesh** [comment]
I'm closing this pr for now since because of this pr, I get an email for each commit you make.

**Pull request closed without merging**